File: tests/verify_qbal_lapsprep.f90
PROGRAM verify_qbal_lapsprep
  USE, INTRINSIC :: iso_fortran_env, ONLY: int32, int64, real64
  USE, INTRINSIC :: ieee_arithmetic, ONLY: ieee_is_finite
  USE netcdf, ONLY: nf90_open,nf90_close,nf90_nowrite,nf90_noerr,nf90_float, &
    nf90_inq_varid,nf90_inquire_variable,nf90_get_var,nf90_get_att,nf90_enotatt, &
    nf90_inq_dimid,nf90_inquire_dimension
  IMPLICIT NONE
  INTEGER, PARAMETER :: nx=6, ny=6, nz=4, nout=nz+1
  INTEGER(int64), PARAMETER :: candidate_size=12_int64+4_int64*(nz+6*nx*ny*nz)
  INTEGER(int64), PARAMETER :: reader_size=12_int64+4_int64*(nout+6*nx*ny*nout)
  INTEGER(int32), PARAMETER :: dimensions3(3)=[6_int32,6_int32,4_int32]
  INTEGER(int32), PARAMETER :: dimensions5(3)=[6_int32,6_int32,5_int32]
  REAL :: candidate_p(nz),u(nx,ny,nz),v(nx,ny,nz),phi(nx,ny,nz)
  REAL :: temperature(nx,ny,nz),q(nx,ny,nz),omega(nx,ny,nz)
  REAL :: pressure(nout),ht(nx,ny,nout),t(nx,ny,nout),mr(nx,ny,nout)
  REAL :: reader_u(nx,ny,nout),reader_v(nx,ny,nout),reader_omega(nx,ny,nout)
  REAL :: expected_ht(nx,ny,nout),expected_t(nx,ny,nout),expected_mr(nx,ny,nout)
  REAL :: expected_u(nx,ny,nout),expected_v(nx,ny,nout),expected_omega(nx,ny,nout)
  REAL :: expected_geometry(8)
  CHARACTER(4096) :: candidate_file,reader_file,wps_file,static_file
  INTEGER :: k,source
  IF (COMMAND_ARGUMENT_COUNT() /= 4) CALL fail('usage: candidate.bin reader_state.bin wps.out static.nest7grid')
  CALL GET_COMMAND_ARGUMENT(1,candidate_file)
  CALL GET_COMMAND_ARGUMENT(2,reader_file)
  CALL GET_COMMAND_ARGUMENT(3,wps_file)
  CALL GET_COMMAND_ARGUMENT(4,static_file)
  CALL read_source_geometry(TRIM(static_file),expected_geometry)
  CALL read_candidate(TRIM(candidate_file),candidate_p,u,v,phi,temperature,q,omega)
  expected_ht=0.0; expected_t=0.0; expected_mr=0.0
  expected_u=0.0; expected_v=0.0; expected_omega=0.0
  DO k=1,nz
    source=nz-k+1
    expected_ht(:,:,k)=phi(:,:,source)/9.80665
    expected_t(:,:,k)=temperature(:,:,source)
    expected_mr(:,:,k)=q(:,:,source)/(1.0-q(:,:,source))
    expected_u(:,:,k)=u(:,:,source)
    expected_v(:,:,k)=v(:,:,source)
    expected_omega(:,:,k)=omega(:,:,source)
  END DO
  expected_ht(:,:,nout)=0.0; expected_t(:,:,nout)=280.0; expected_mr(:,:,nout)=8.0/1000.0
  expected_u(:,:,nout)=2.0; expected_v(:,:,nout)=-2.0; expected_omega(:,:,nout)=0.0
  CALL read_reader(TRIM(reader_file),pressure,ht,t,mr,reader_u,reader_v,reader_omega)
  CALL check_vector(pressure,[700.0,800.0,900.0,1000.0,2001.0],'reader pressure')
  CALL check_vector(RESHAPE(ht,[SIZE(ht)]),RESHAPE(expected_ht,[SIZE(ht)]),'reader ht')
  CALL check_vector(RESHAPE(t,[SIZE(t)]),RESHAPE(expected_t,[SIZE(t)]),'reader t')
  CALL check_vector(RESHAPE(mr,[SIZE(mr)]),RESHAPE(expected_mr,[SIZE(mr)]),'reader mr')
  CALL check_vector(RESHAPE(reader_u,[SIZE(reader_u)]),RESHAPE(expected_u,[SIZE(reader_u)]),'reader u')
  CALL check_vector(RESHAPE(reader_v,[SIZE(reader_v)]),RESHAPE(expected_v,[SIZE(reader_v)]),'reader v')
  CALL check_vector(RESHAPE(reader_omega,[SIZE(reader_omega)]), &
                    RESHAPE(expected_omega,[SIZE(reader_omega)]),'reader omega')
  CALL check_wps(TRIM(wps_file),expected_ht,expected_t,expected_mr,expected_u,expected_v)
  WRITE(*,'(A)') 'PASS_SCOPED'
CONTAINS
  SUBROUTINE fail(message)
    CHARACTER(*), INTENT(IN) :: message
    WRITE(*,'(A)') 'FAIL: '//TRIM(message)
    ERROR STOP 1
  END SUBROUTINE fail
  LOGICAL FUNCTION same_bits(a,b)
    REAL, INTENT(IN) :: a,b
    INTEGER(int32) :: ia,ib
    ia=TRANSFER(a,ia)
    ib=TRANSFER(b,ib)
    same_bits=(ia == ib)
  END FUNCTION same_bits
  SUBROUTINE check_vector(actual,expected,label)
    REAL, INTENT(IN) :: actual(:),expected(:)
    CHARACTER(*), INTENT(IN) :: label
    INTEGER :: i
    DO i=1,SIZE(actual)
      IF (.NOT.ieee_is_finite(actual(i)) .OR. .NOT.ieee_is_finite(expected(i))) &
        CALL fail(TRIM(label)//': nonfinite')
      IF (.NOT.same_bits(actual(i),expected(i))) CALL fail(TRIM(label)//': bits differ')
    END DO
  END SUBROUTINE check_vector
  SUBROUTINE read_source_geometry(name,geometry)
    CHARACTER(*), INTENT(IN) :: name
    REAL, INTENT(OUT) :: geometry(8)
    CHARACTER(6), PARAMETER :: variables(7)=[character(6) :: 'lat','lon','Dx','Dy','LoV','Latin1','Latin2']
    CHARACTER(13), PARAMETER :: missing_attributes(2)=[character(13) :: '_FillValue','missing_value']
    INTEGER :: ncid,vid,status,xtype,rank,i,j,dimid,length
    INTEGER :: start(4)
    REAL(real64) :: missing
    CHARACTER(132) :: projection

    status=nf90_open(name,nf90_nowrite,ncid)
    IF (status /= nf90_noerr) CALL fail('source static open failed')
    status=nf90_inq_dimid(ncid,'x',dimid)
    IF (status == nf90_noerr) status=nf90_inquire_dimension(ncid,dimid,len=length)
    IF (status /= nf90_noerr) CALL fail('source static x dimension missing')
    IF (length /= nx) CALL fail('source static x dimension mismatch')
    status=nf90_inq_dimid(ncid,'y',dimid)
    IF (status == nf90_noerr) status=nf90_inquire_dimension(ncid,dimid,len=length)
    IF (status /= nf90_noerr) CALL fail('source static y dimension missing')
    IF (length /= ny) CALL fail('source static y dimension mismatch')
    status=nf90_inq_varid(ncid,'grid_type',vid)
    IF (status == nf90_noerr) status=nf90_get_var(ncid,vid,projection)
    IF (status /= nf90_noerr) CALL fail('source static projection read failed')
    IF (TRIM(projection) /= 'secant lambert conformal') CALL fail('source static projection unsupported')
    start=1
    DO i=1,SIZE(variables)
      status=nf90_inq_varid(ncid,TRIM(variables(i)),vid)
      IF (status == nf90_noerr) status=nf90_inquire_variable(ncid,vid,xtype=xtype,ndims=rank)
      IF (status /= nf90_noerr) CALL fail('source static variable missing: '//TRIM(variables(i)))
      IF (xtype /= nf90_float .OR. rank < 1 .OR. rank > 4) CALL fail('source static layout mismatch')
      status=nf90_get_var(ncid,vid,geometry(i),start=start(:rank))
      IF (status /= nf90_noerr) CALL fail('source static value read failed')
      IF (.NOT.ieee_is_finite(geometry(i))) CALL fail('source geometry: nonfinite')
      DO j=1,SIZE(missing_attributes)
        status=nf90_get_att(ncid,vid,TRIM(missing_attributes(j)),missing)
        IF (status == nf90_noerr) THEN
          IF (ieee_is_finite(missing)) THEN
            IF (REAL(geometry(i),real64) == missing) CALL fail('source geometry: missing value')
          END IF
        ELSE IF (status /= nf90_enotatt) THEN
          CALL fail('source geometry: invalid missing attribute')
        END IF
      END DO
    END DO
    status=nf90_close(ncid)
    IF (status /= nf90_noerr) CALL fail('source static close failed')
    ! Static navigation is in degrees and meters; WPS spacing/radius is in km.
    IF (geometry(1) > 270.) geometry(1)=geometry(1)-360.
    IF (geometry(2) > 180.) geometry(2)=geometry(2)-360.
    geometry(3:4)=geometry(3:4)/1000.
    IF (geometry(5) > 180.) geometry(5)=geometry(5)-360.
    geometry(8)=6371.229  ! Declared Earth radius in the maintained WPS writer.
    IF (ANY(geometry(3:4) <= 0.)) CALL fail('source geometry: nonpositive scale')
  END SUBROUTINE read_source_geometry
  SUBROUTINE read_candidate(name,p,u,v,phi,temp,moisture,om)
    CHARACTER(*), INTENT(IN) :: name
    REAL, INTENT(OUT) :: p(:),u(:,:,:),v(:,:,:),phi(:,:,:),temp(:,:,:),moisture(:,:,:),om(:,:,:)
    INTEGER(int32) :: dimensions(3)
    INTEGER(int64) :: file_size
    INTEGER :: unit,ios
    LOGICAL :: exists
    INQUIRE(FILE=name,EXIST=exists,SIZE=file_size)
    IF (.NOT.exists .OR. file_size /= candidate_size) CALL fail('candidate file size mismatch')
    OPEN(NEWUNIT=unit,FILE=name,FORM='UNFORMATTED',ACCESS='STREAM',STATUS='OLD', &
         ACTION='READ',CONVERT='LITTLE_ENDIAN',IOSTAT=ios)
    IF (ios /= 0) CALL fail('candidate open failed')
    READ(unit,IOSTAT=ios) dimensions
    IF (ios /= 0 .OR. ANY(dimensions /= dimensions3)) CALL fail('candidate dimensions mismatch')
    READ(unit,IOSTAT=ios) p,u,v,phi,temp,moisture,om
    IF (ios /= 0) CALL fail('candidate fields read failed')
    CLOSE(unit,IOSTAT=ios)
    IF (ios /= 0) CALL fail('candidate close failed')
    CALL check_vector(p,[100000.0,90000.0,80000.0,70000.0],'candidate pressure')
  END SUBROUTINE read_candidate
  SUBROUTINE read_reader(name,p,ht,t,mr,u,v,om)
    CHARACTER(*), INTENT(IN) :: name
    REAL, INTENT(OUT) :: p(:),ht(:,:,:),t(:,:,:),mr(:,:,:),u(:,:,:),v(:,:,:),om(:,:,:)
    INTEGER(int32) :: dimensions(3)
    INTEGER(int64) :: file_size
    INTEGER :: unit,ios
    LOGICAL :: exists
    INQUIRE(FILE=name,EXIST=exists,SIZE=file_size)
    IF (.NOT.exists .OR. file_size /= reader_size) CALL fail('reader state file size mismatch')
    OPEN(NEWUNIT=unit,FILE=name,FORM='UNFORMATTED',ACCESS='STREAM',STATUS='OLD', &
         ACTION='READ',CONVERT='LITTLE_ENDIAN',IOSTAT=ios)
    IF (ios /= 0) CALL fail('reader state open failed')
    READ(unit,IOSTAT=ios) dimensions
    IF (ios /= 0 .OR. ANY(dimensions /= dimensions5)) CALL fail('reader state dimensions mismatch')
    READ(unit,IOSTAT=ios) p,ht,t,mr,u,v,om
    IF (ios /= 0) CALL fail('reader state fields read failed')
    CLOSE(unit,IOSTAT=ios)
    IF (ios /= 0) CALL fail('reader state close failed')
  END SUBROUTINE read_reader
  INTEGER FUNCTION selected_field(name)
    CHARACTER(*), INTENT(IN) :: name
    SELECT CASE (TRIM(name))
    CASE ('UU'); selected_field=1
    CASE ('VV'); selected_field=2
    CASE ('TT'); selected_field=3
    CASE ('HGT'); selected_field=4
    CASE ('QV'); selected_field=5
    CASE DEFAULT; selected_field=0
    END SELECT
  END FUNCTION selected_field
  SUBROUTINE check_wps(name,expected_ht,expected_t,expected_mr,expected_u,expected_v)
    CHARACTER(*), INTENT(IN) :: name
    REAL, INTENT(IN) :: expected_ht(:,:,:),expected_t(:,:,:),expected_mr(:,:,:)
    REAL, INTENT(IN) :: expected_u(:,:,:),expected_v(:,:,:)
    REAL, PARAMETER :: levels(5)=[70000.0,80000.0,90000.0,100000.0,200100.0]
    CHARACTER(24), PARAMETER :: expected_time='2023-05-18_03:33:20.0000'
    CHARACTER(4), PARAMETER :: names(5)=(/'UU  ','VV  ','TT  ','HGT ','QV  '/)
    CHARACTER(25), PARAMETER :: units_expected(5)=[character(25) :: 'm s{-1}','m s{-1}','K','m','kg kg{-1}']
    LOGICAL :: seen(5,5),exists
    CHARACTER(24) :: hdate
    CHARACTER(32) :: source_name
    CHARACTER(9) :: field
    CHARACTER(25) :: units
    CHARACTER(46) :: description
    CHARACTER(8) :: knownloc
    REAL :: xfcst,level,la1,lo1,dx,dy,lov,latin1,latin2,earth_radius
    REAL :: slab(nx,ny)
    INTEGER(int32) :: wind_raw
    INTEGER :: unit,ios,version,nx_file,ny_file,llflag,wind_field
    INTEGER :: i,level_index
    INTEGER :: detect_unit
    CHARACTER(4) :: marker
    CHARACTER(13) :: byte_order
    CHARACTER(80) :: label,level_text
    seen=.FALSE.
    INQUIRE(FILE=name,EXIST=exists)
    IF (.NOT.exists) CALL fail('WPS file missing')
    OPEN(NEWUNIT=detect_unit,FILE=name,FORM='UNFORMATTED',ACCESS='STREAM',STATUS='OLD', &
         ACTION='READ',IOSTAT=ios)
    IF (ios /= 0) CALL fail('WPS open failed')
    READ(detect_unit,IOSTAT=ios) marker
    CLOSE(detect_unit)
    IF (ios /= 0) CALL fail('WPS record marker read failed')
    IF (IACHAR(marker(1:1)) == 0 .AND. IACHAR(marker(2:2)) == 0 .AND. &
        IACHAR(marker(3:3)) == 0 .AND. IACHAR(marker(4:4)) == 4) THEN
      byte_order='BIG_ENDIAN'
    ELSE IF (IACHAR(marker(1:1)) == 4 .AND. IACHAR(marker(2:2)) == 0 .AND. &
             IACHAR(marker(3:3)) == 0 .AND. IACHAR(marker(4:4)) == 0) THEN
      byte_order='LITTLE_ENDIAN'
    ELSE
      CALL fail('WPS record marker invalid')
    END IF
    OPEN(NEWUNIT=unit,FILE=name,FORM='UNFORMATTED',ACCESS='SEQUENTIAL',STATUS='OLD', &
         ACTION='READ',CONVERT=TRIM(byte_order),IOSTAT=ios)
    IF (ios /= 0) CALL fail('WPS open failed')
    DO
      READ(unit,IOSTAT=ios) version
      IF (ios < 0) EXIT
      IF (ios /= 0 .OR. version /= 5) CALL fail('WPS version mismatch')
      READ(unit,IOSTAT=ios) hdate,xfcst,source_name,field,units,description,level, &
        nx_file,ny_file,llflag
      IF (ios /= 0) CALL fail('WPS metadata read failed')
      CALL check_vector([xfcst],[0.],'WPS XFCST')
      IF (hdate /= expected_time) CALL fail('WPS time mismatch')
      IF (nx_file /= nx .OR. ny_file /= ny) CALL fail('WPS shape mismatch')
      IF (llflag /= 3) CALL fail('WPS projection mismatch')
      READ(unit,IOSTAT=ios) knownloc,la1,lo1,dx,dy,lov,latin1,latin2,earth_radius
      IF (ios /= 0) CALL fail('WPS projection read failed')
      IF (knownloc /= 'SWCORNER') CALL fail('WPS KNOWNLOC mismatch')
      IF (.NOT.ALL(ieee_is_finite([la1,lo1,dx,dy,lov,latin1,latin2,earth_radius]))) &
        CALL fail('WPS geometry: nonfinite')
      IF (dx <= 0. .OR. dy <= 0. .OR. earth_radius <= 0.) CALL fail('WPS geometry: nonpositive scale')
      CALL check_vector([la1,lo1,dx,dy,lov,latin1,latin2,earth_radius],expected_geometry,'WPS geometry')
      READ(unit,IOSTAT=ios) wind_raw
      IF (ios /= 0) CALL fail('WPS wind metadata read failed')
      IF (wind_raw /= -1 .AND. wind_raw /= 0 .AND. wind_raw /= 1) &
        CALL fail('WPS wind metadata invalid')
      READ(unit,IOSTAT=ios) slab
      IF (ios /= 0) CALL fail('WPS slab read failed')
      IF (TRIM(field) == 'W' .OR. TRIM(field) == 'WW' .OR. TRIM(field) == 'OM' .OR. &
          TRIM(field) == 'OMEGA') CALL fail('unexpected WPS omega field')
      wind_field=selected_field(field)
      IF (wind_field == 0) CYCLE
      IF (wind_raw == 0) CALL fail('WPS '//TRIM(field)//': wind basis mismatch')
      IF (TRIM(units) /= TRIM(units_expected(wind_field))) &
        CALL fail('WPS '//TRIM(field)//': metadata mismatch')
      level_index=0
      DO i=1,5
        IF (same_bits(level,levels(i))) level_index=i
      END DO
      IF (level_index == 0) CALL fail('WPS '//TRIM(field)//': pressure inventory mismatch')
      IF (seen(wind_field,level_index)) CALL fail('WPS '//TRIM(field)//': duplicate record')
      seen(wind_field,level_index)=.TRUE.
      WRITE(level_text,'(F6.1)') levels(level_index)/100.0
      label='WPS '//TRIM(field)//' '//ADJUSTL(level_text)
      SELECT CASE (wind_field)
      CASE (1); CALL check_vector(RESHAPE(slab,[SIZE(slab)]), &
          RESHAPE(expected_u(:,:,level_index),[SIZE(slab)]),label)
      CASE (2); CALL check_vector(RESHAPE(slab,[SIZE(slab)]), &
          RESHAPE(expected_v(:,:,level_index),[SIZE(slab)]),label)
      CASE (3); CALL check_vector(RESHAPE(slab,[SIZE(slab)]), &
          RESHAPE(expected_t(:,:,level_index),[SIZE(slab)]),label)
      CASE (4); CALL check_vector(RESHAPE(slab,[SIZE(slab)]), &
          RESHAPE(expected_ht(:,:,level_index),[SIZE(slab)]),label)
      CASE (5); CALL check_vector(RESHAPE(slab,[SIZE(slab)]), &
          RESHAPE(expected_mr(:,:,level_index),[SIZE(slab)]),label)
      END SELECT
    END DO
    CLOSE(unit,IOSTAT=ios)
    IF (ios /= 0) CALL fail('WPS close failed')
    DO wind_field=1,5
      DO level_index=1,5
        IF (.NOT.seen(wind_field,level_index)) &
          CALL fail('WPS '//TRIM(names(wind_field))//': pressure inventory mismatch')
      END DO
    END DO
  END SUBROUTINE check_wps
END PROGRAM verify_qbal_lapsprep

File: tests/verify_qbal_lapsprep.py
#!/usr/bin/env python3
"""Scoped writer -> actual cold LAPSPREP caller/WPS handoff oracle.

The CDF test hook observes the caller's arrays; it is not a native model writer.
Cold LAPSPREP retains pressure-level omega in w; its surface slot is LSX VV
(m/s). WPS has no omega slab.
"""

from datetime import datetime, timedelta, timezone
import hashlib
import json
from pathlib import Path
import shutil
import struct
import sys

import netCDF4
import numpy as np

from verify_qbal_writer import require, verify as verify_writer

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "tools"))
from compare_baseline import read_wps, _records

I4TIME = 2000000000
VALID_TIME = datetime(1960, 1, 1, tzinfo=timezone.utc) + timedelta(seconds=I4TIME)
STAMP = VALID_TIME.strftime("%y%j%H%M")
WPS_TIME = VALID_TIME.replace(second=0).strftime("%Y-%m-%d_%H:%M:%S.0000")


def sha256(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()


def preflight(variant):
    """Validate the reviewed products plus the ancillary file used for pressure."""
    verify_writer(variant / "candidate.bin", variant / "out", I4TIME)
    for extension in ("lt1", "lw3", "lh3", "lq3"):
        directory = variant / "out" / "balance" / extension
        require([p.name for p in directory.glob("*." + extension)] == [f"{STAMP}.{extension}"],
                f"{extension}: filename/time mismatch")
    with netCDF4.Dataset(variant / "out/balance/lh3" / f"{STAMP}.lh3") as ds:
        require(tuple(len(ds.dimensions[n]) for n in ("record", "z", "y", "x")) == (1, 4, 6, 6),
                "lh3: dimensions mismatch")
        for name, expected in (("level", [700., 800., 900., 1000.]),
                               ("reftime", [I4TIME - 315619200]),
                               ("valtime", [I4TIME - 315619200])):
            variable = ds[name]
            expected_units = "hectopascals" if name == "level" else "seconds since (1970-1-1 00:00:00.0)"
            require(variable.units == expected_units, f"lh3: {name}: units mismatch")
            values = variable[:]
            require(not np.ma.getmaskarray(values).any(), f"lh3: {name}: unexpected missing mask")
            require(np.isfinite(np.asarray(values)).all() and np.array_equal(values, expected),
                    f"lh3: {name}: coordinate mismatch")
        require(ds["rhl"].dimensions == ("record", "z", "y", "x")
                and ds["rhl"].units == "percent", "lh3: RH layout/units mismatch")
        rh = ds["rhl"][:]
        require(not np.ma.getmaskarray(rh).any() and np.all(rh == 50.), "lh3: invalid fixture RH")
        inventory = ds["rhl_fcinv"][:]
        require(not np.ma.getmaskarray(inventory).any() and np.all(inventory == 1),
                "lh3: incomplete inventory")


def stage(variant, root):
    preflight(variant)
    root.mkdir(parents=True, exist_ok=False)
    (root / "static").mkdir()
    shutil.copytree(variant / "out/balance", root / "lapsprd/balance")
    (root / "lapsprd/lsx").mkdir()
    with netCDF4.Dataset(root / "lapsprd/lsx" / f"{STAMP}.lsx", "w", format="NETCDF3_CLASSIC") as ds:
        for name, size in (("record", 1), ("z", 1), ("y", 6), ("x", 6)):
            ds.createDimension(name, size)
        values = {"u": 2., "v": -2., "t": 280., "rh": 50., "tgd": 285.,
                  "ps": 101000., "msl": 101000., "mr": 8., "vv": 0.}
        for name, value in values.items():
            ds.createVariable(name, "f4", ("record", "z", "y", "x"))[:] = value
    with netCDF4.Dataset(root / "static/static.nest7grid", "w", format="NETCDF3_CLASSIC") as ds:
        for name, size in (("record", 1), ("z", 1), ("y", 6), ("x", 6), ("nav", 1), ("namelen", 132)):
            ds.createDimension(name, size)
        for name, value in {"Dx": 10000., "Dy": 10000., "LoV": 127., "Latin1": 45., "Latin2": 45.}.items():
            ds.createVariable(name, "f4", ("nav",))[:] = value
        ds.createVariable("grid_type", "S1", ("nav", "namelen"))[:] = np.frombuffer(
            b"secant lambert conformal".ljust(132, b" "), dtype="S1")
        for name, value in {"lat": np.broadcast_to(np.linspace(45., 45.5, 6)[:, None], (6, 6)),
                            "lon": np.broadcast_to(np.linspace(127., 127.5, 6), (6, 6)), "avg": 0.}.items():
            ds.createVariable(name, "f4", ("record", "z", "y", "x"))[:] = value
    (root / "static/lapsprep.nl").write_text(
        "&lapsprep_nl\n HOTSTART=.false., BALANCE=.true., OUTPUT_FORMAT='wps','cdf',\n"
        " GRID_SCALE='NONE', HYDRO_MODE='CONSERVATIVE', CAP_POLICY='KEEP',\n"
        " WIND_COORDINATE='GRID_RELATIVE', MAKE_SFC_UV=.false.,\n"
        " ENFORCE_FIELD_CONTRACTS=.true., WPS_OUTPUT_VAPOR=.true., LWC2VAPOR_THRESH=0.0,\n/\n")
    products = {str(p.relative_to(root)): sha256(p) for p in (root / "lapsprd").rglob("*") if p.is_file()}
    products.update({str(p.relative_to(root)): sha256(p) for p in (root / "static").iterdir()})
    (root / "staged_inputs.json").write_text(json.dumps(products, indent=2) + "\n")
    return STAMP


def same_bits(actual, expected, name):
    require(actual.shape == expected.shape and np.isfinite(actual).all(), f"{name}: shape/nonfinite")
    require(np.array_equal(np.asarray(actual, dtype="<f4").view("<u4"),
                           np.asarray(expected, dtype="<f4").view("<u4")), f"{name}: bits differ")


def verify(variant, root, wps_path):
    preflight(variant)
    for relative, digest in json.loads((root / "staged_inputs.json").read_text()).items():
        require(sha256(root / relative) == digest, f"staged input changed: {relative}")
    for source in (variant / "out/balance").rglob("*"):
        if source.is_file():
            target = root / "lapsprd/balance" / source.relative_to(variant / "out/balance")
            require(sha256(source) == sha256(target), "writer/staged bytes differ")
    raw = (variant / "candidate.bin").read_bytes()
    u, v, phi, temperature, q, omega = np.frombuffer(raw, dtype="<f4", offset=28).reshape(6, 4, 6, 6)
    expected = {"ht": (phi / np.float32(9.80665))[::-1], "t": temperature[::-1],
                "mr": (q / (np.float32(1.) - q))[::-1], "u": u[::-1], "v": v[::-1], "omega": omega[::-1]}
    state = (root / "reader_state.bin").read_bytes()
    require(len(state) == 12 + 4 * (5 + 6 * 5 * 6 * 6), "reader state length mismatch")
    require(np.array_equal(np.frombuffer(state, dtype="<i4", count=3), [6, 6, 5]), "reader state dimensions")
    pressure = np.frombuffer(state, dtype="<f4", count=5, offset=12)
    same_bits(pressure, np.array([700, 800, 900, 1000, 2001], dtype="<f4"), "reader pressure")
    arrays = np.frombuffer(state, dtype="<f4", offset=32).reshape(6, 5, 6, 6)
    surfaces = {"ht": 0., "t": 280., "mr": np.float32(8.) / np.float32(1000.), "u": 2., "v": -2., "omega": 0.}
    for (name, values), actual in zip(expected.items(), arrays):
        same_bits(actual[:4], values, f"reader {name}")
        same_bits(actual[4], np.full((6, 6), surfaces[name], dtype="<f4"), f"surface {'VV (m/s)' if name == 'omega' else name}")
    wps = read_wps(wps_path)
    require({key[2] for key in wps} == {WPS_TIME}, "WPS time mismatch")
    require(not any(key[0] in {"W", "WW", "OM", "OMEGA"} for key in wps), "unexpected WPS omega field")
    mapping = {"UU": ("u", "m s{-1}"), "VV": ("v", "m s{-1}"), "TT": ("t", "K"),
               "HGT": ("ht", "m"), "QV": ("mr", "kg kg{-1}")}
    for field, (name, unit) in mapping.items():
        require({key[1] for key in wps if key[0] == field} == {70000., 80000., 90000., 100000., 200100.},
                f"WPS {field}: pressure inventory mismatch")
        for k, p in enumerate(pressure):
            units, shape, slab, metadata = wps[(field, float(p * 100), WPS_TIME)]
            require(units == unit and tuple(shape) == (6, 6), f"WPS {field}: metadata mismatch")
            require(metadata["wind_grid_relative"], f"WPS {field}: wind basis mismatch")
            values = expected[name][k] if k < 4 else np.full((6, 6), surfaces[name], dtype="<f4")
            same_bits(slab.reshape(6, 6), values, f"WPS {field} {p}")
    return {"status": "PASS_SCOPED", "scope": "actual cold LAPSPREP caller arrays and five WPS pressure fields",
            "candidate_sha256": sha256(variant / "candidate.bin"),
            "reader_fields": list(expected), "wps_fields": list(mapping),
            "transformations": {"height": "float32 PHI/9.80665", "vapor": "float32 q/(1-q)",
                                "pressure": "ascending hPa in caller; Pa in WPS", "omega": "pressure slots retain Pa/s; surface slot is LSX VV (m/s); absent from WPS"},
            "time": {"file": VALID_TIME.isoformat(), "wps": WPS_TIME, "lost_seconds": VALID_TIME.second,
                     "status": "OPEN: legacy filename/WPS minute precision loses 20 seconds"},
            "open": ["time-preserving native handoff", "terrain/missing-mask correspondence", "native seed/startup/halo consumed state",
                     "hotstart omega-to-W and thermodynamics", "full mass closure", "forecast effects"],
            "limitations": ["synthetic all-valid 6x6x4; synthetic LSX/static/RH",
                            "test-only CDF hook is not a native writer; no model startup executed",
                            "mask/unit/time preflight belongs to Python gate, not legacy NCVGT"]}


def corrupt_wps_qv(path):
    records = list(_records(path))
    for index in range(0, len(records), 5):
        if records[index + 1][1][60:69].decode("ascii").strip() == "QV":
            endian, slab = records[index + 4]
            slab = bytearray(slab)
            np.frombuffer(slab, dtype=endian + "f4", count=1)[0] += 1.
            records[index + 4] = (endian, slab)
            break
    else:
        raise ValueError("WPS QV control target absent")
    with path.open("wb") as stream:
        for endian, payload in records:
            marker = struct.pack(endian + "i", len(payload))
            stream.write(marker + payload + marker)


def controls(variant, root, wps_path):
    """Corrupt copied handoff artifacts and require the designated oracle error."""
    verify(variant, root, wps_path)
    negative = root.parent / (root.name + "-negative")
    negative.mkdir(exist_ok=False)
    results = []
    names = ["ht", "t", "mr", "u", "v", "omega"]
    for index, name in enumerate([*names, "pressure", "wps_time", "wps_qv"]):
        target = negative / name
        shutil.copytree(root, target, ignore=shutil.ignore_patterns("build"))
        output = target / "control.wps"
        shutil.copyfile(wps_path, output)
        if name == "wps_qv":
            corrupt_wps_qv(output)
            expected = "WPS QV 700.0: bits differ"
        elif name == "wps_time":
            raw = output.read_bytes()
            require(WPS_TIME.encode() in raw, "WPS time control target absent")
            output.write_bytes(raw.replace(WPS_TIME.encode(), WPS_TIME.replace("03:33:", "03:34:").encode()))
            expected = "WPS time mismatch"
        else:
            path = target / "reader_state.bin"
            raw = bytearray(path.read_bytes())
            offset = 12 if name == "pressure" else 32 + index * 5 * 6 * 6 * 4
            value = np.frombuffer(raw, dtype="<f4", count=1, offset=offset)
            value[0] += np.float32(1.)
            path.write_bytes(raw)
            expected = "reader pressure: bits differ" if name == "pressure" else f"reader {name}: bits differ"
        try:
            verify(variant, target, output)
        except ValueError as error:
            require(str(error) == expected, f"{name}: wrong rejection: {error}")
            results.append({"case": name, "status": "REJECTED", "reason": str(error)})
        else:
            raise ValueError(f"undetected handoff corruption: {name}")
    for name in ("level", "reftime", "valtime"):
        target = negative / ("lh3_missing_" + name)
        (target / "out").mkdir(parents=True)
        shutil.copyfile(variant / "candidate.bin", target / "candidate.bin")
        shutil.copytree(variant / "out/balance", target / "out/balance")
        with netCDF4.Dataset(target / "out/balance/lh3" / f"{STAMP}.lh3", "r+") as ds:
            ds[name].setncattr("missing_value", ds[name][:].flat[0].item())
        try:
            preflight(target)
        except ValueError as error:
            require(str(error) == f"lh3: {name}: unexpected missing mask", f"wrong mask rejection: {error}")
            results.append({"case": "lh3_missing_" + name, "status": "REJECTED", "reason": str(error)})
        else:
            raise ValueError(f"undetected ancillary mask: {name}")
    return results


if __name__ == "__main__":
    command, *arguments = sys.argv[1:]
    paths = list(map(Path, arguments))
    if command == "stage":
        print(stage(*paths))
    elif command == "verify":
        print(json.dumps(verify(*paths), indent=2))
    elif command == "controls":
        print(json.dumps(controls(*paths), indent=2))
    else:
        raise SystemExit("usage: verify_qbal_lapsprep.py stage VARIANT ROOT | verify VARIANT ROOT WPS")
